explainer-datamodel.md

```diff
@@ -1,236 +1,329 @@
+individual text objects (in the first parameter).
 
-### Cascading of values from FormattedText to text runs
+### Specifying constraints
 
-CSS properties applied to the style map from the FormattedText that are specified to inherit
-from parent to child will do so from the FormattedText to it's text run child objects.
+The final input to `format` are any constraints that should be applied to the formatted text.
+The primary constraints of interest to the API are the width and height of the containing block:
+how many CSS pixels are available for text layout.
 
-## Simple Example
+The layout flow will follow any writing-mode direction specified in the meta object, so in
+`horizontal-lr` (e.g., English) a width constraint on the containing block will cause line 
+wrapping in the inline direction; overflow will occur in the vertical direction. For specified 
+writing-modes such as `vertical-rl` (e.g., Chinese) a width constraint on the containing block 
+will only affect where overflow occurs--the line will not wrap because it is assumed to have 
+infinite vertical space to layout. To cause wrapping with `vertical-rl` writing modes, specify
+a height constraint instead.
 
-In this example, we create a `FormattedText` object and its text runs that we eventually 
-want to render with the text "The quick **brown** fox jumps over the lazy dog" where the
-word "brown" is colored brown and bold. In the [rendering explainer](explainer-rendering.md) we
-add layout constraints and render it to a canvas; at present, while it resides in the data model
-it has no associated layout.
+Constraints are specified on a constraint object using a `width` or `height` property:
 
 ```js
-// Collect text into a FormattedText object
-let formattedText = new FormattedText( "The quick ", "brown", " fox jumps over the lazy dog" );
-formattedText.textruns[1].setStyle( "color: brown; font-weight: bold" );
+// Wrap any text that exceeds 150 pixels
+FormattedText.format( "The quick brown fox jumps over the lazy dog.", null, { width: 150 } );
 ```
 
-The above is the semantic equivalent to the following declarative HTML markup:
+`width` and `height` are `unsigned long` values. Omitted values are assumed to be infinite.
+
+## Comparison to HTML
+
+`format` is used to drive the web platform's layout engine, but using the JavaScript-based data
+model described above. Therefore, the output of `format` should be the equivalent to what can be
+already performed in HTML using simple elements like `div` and `span`. The following two expressions
+are functionally equivalent, with the exception that the result of `format` has not been rendered
+(and thus can't be visualized yet):
+
+```js
+FormattedText.format( [ "The quick ",
+                        { text: "brown", style: "color:brown;font-weight:bold" },
+                        " fox jumps over the lazy dog"
+                      ], null, { width: 150 } );
```

Comment:
Have the implementors evaluated the performance implications of this API for the case when the text contents remain the same while the constraints change rapidly? In this case the application would be passing the same array of text runs but pass a new `width`/`height`. If caching could provide performance benefits, then the text engine would also have to perform the following extra computation to produce a cache hit:

* Make a defensive copy of the previously formatted text.
* Fully traverse the newly input and compare to the previous snapshot.

The engine has to do these things because the text is expressed using plain _mutable_ JavaScript objects (arrays, dictionaries). It's not possible to tell if something inside changed without the extra work.

A common use-case is resizing a window, which transitively resizes text inside. While all the strings and styles remain the same the value of `width` would change on every frame for multiple paragraphs of text in the UI. Some apps support resizing within the app, such as columns and rows of a spreadsheet.

Reply:
This is a very good use case to consider and to test for. My expectation at this early stage is that this case would be sub-optimal and that no caching/comparing of deltas between iterations of `format` are maintained.

These were all good reasons to have a retained data model supported by the platform--changes to the data model could trigger invalidation (or not) and make formatting faster. However, we're making the trade-off with this new approach to remove any retained platform data model and rely on JS strings, and as a consequence, this issue is now very much a potential concern.

There may be an opportunity for a new feature here: given an existing retained metrics object, allow it to be "re-calculated" based only on changing the constraints--no changes to style or text content. Since no new input is needed except for the new constraint it seems possible to make a fast adjustment to the existing formatted metrics and output (or in-place update) new metrics. 

Logged as Issue #45

Reply:
Can continue discussion on issue #45